// myconf/lib/myconf.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var _homeDir = require('home-dir');

var _homeDir2 = _interopRequireDefault(_homeDir);

var _utils = require('./utils');

var _utils2 = _interopRequireDefault(_utils);

require('babel/polyfill');

var _path2 = require('path');

var _path3 = _interopRequireDefault(_path2);

var dotFileRe = /^(\.[^\/]*)$/;

var Config = (function () {
  function Config(filename) {
    _classCallCheck(this, Config);

    var validFilename = dotFileRe.test(filename);
    if (!validFilename) {
      return console.error('the filename \'' + filename + '\' is not valid');
    }
    this.filename = filename;
    this.filepath = (0, _homeDir2['default'])(filename);
    this.fileparser = 'json';
  }

  _createClass(Config, [{
    key: 'path',
    value: function path() {
      var filepath = arguments.length <= 0 || arguments[0] === undefined ? (0, _homeDir2['default'])() : arguments[0];

      this.filepath = _path3['default'].resolve(filepath + ('/' + this.filename));
      return this;
    }
  }, {
    key: 'parser',
    value: function parser() {
      var _parser = arguments.length <= 0 || arguments[0] === undefined ? 'json' : arguments[0];

      this.fileparser = _parser;
      return this;
    }
  }, {
    key: 'get',
    value: function get(key) {
      var _this = this;

      return new Promise(function callee$2$0(resolve, reject) {
        var filedata;
        return regeneratorRuntime.async(function callee$2$0$(context$3$0) {
          while (1) switch (context$3$0.prev = context$3$0.next) {
            case 0:
              context$3$0.next = 2;
              return regeneratorRuntime.awrap(_utils2['default'].readFile(this.filepath, this.fileparser));

            case 2:
              filedata = context$3$0.sent;

              if (!key) {
                context$3$0.next = 5;
                break;
              }

              return context$3$0.abrupt('return', resolve(filedata && filedata[key]));

            case 5:
              return context$3$0.abrupt('return', resolve(filedata));

            case 6:
            case 'end':
              return context$3$0.stop();
          }
        }, null, _this);
      });
    }
  }, {
    key: 'set',
    value: function set(key, value) {
      var _this2 = this;

      return new Promise(function callee$2$0(resolve, reject) {
        var filedata, prop;
        return regeneratorRuntime.async(function callee$2$0$(context$3$0) {
          while (1) switch (context$3$0.prev = context$3$0.next) {
            case 0:
              context$3$0.next = 2;
              return regeneratorRuntime.awrap(_utils2['default'].readFile(this.filepath, this.fileparser));

            case 2:
              filedata = context$3$0.sent;

              filedata = filedata ? filedata : {};
              if (typeof key === 'object') {
                for (prop in key) {
                  filedata[prop] = key[prop];
                }
              } else if (typeof key === 'string') {
                filedata[key] = value;
              }
              context$3$0.prev = 5;
              context$3$0.next = 8;
              return regeneratorRuntime.awrap(_utils2['default'].saveFile(this.filepath, filedata, this.fileparser));

            case 8:
              resolve(filedata);
              context$3$0.next = 14;
              break;

            case 11:
              context$3$0.prev = 11;
              context$3$0.t0 = context$3$0['catch'](5);

              console.error(context$3$0.t0);

            case 14:
            case 'end':
              return context$3$0.stop();
          }
        }, null, _this2, [[5, 11]]);
      });
    }
  }, {
    key: 'save',
    value: function save(filedata) {
      var _this3 = this;

      return new Promise(function callee$2$0(resolve, reject) {
        return regeneratorRuntime.async(function callee$2$0$(context$3$0) {
          while (1) switch (context$3$0.prev = context$3$0.next) {
            case 0:
              context$3$0.prev = 0;
              context$3$0.next = 3;
              return regeneratorRuntime.awrap(_utils2['default'].saveFile(this.filepath, filedata, this.fileparser));

            case 3:
              resolve(filedata);
              context$3$0.next = 9;
              break;

            case 6:
              context$3$0.prev = 6;
              context$3$0.t0 = context$3$0['catch'](0);

              console.error(context$3$0.t0);

            case 9:
            case 'end':
              return context$3$0.stop();
          }
        }, null, _this3, [[0, 6]]);
      });
    }
  }]);

  return Config;
})();

exports['default'] = Config;
module.exports = exports['default'];
